packages/webpack/lib/middleware/render.js

```diff
@@ -1,41 +1,79 @@
 'use strict';
 
 const { join } = require('path');
+const EventEmitter = require('events');
 
 const webpack = require('webpack');
 const MemoryFS = require('memory-fs');
 
 const sourceMapSupport = require('source-map-support');
 const requireFromString = require('require-from-string');
 
-const { Resolvable } = require('../utils/resolvable');
-
 module.exports = exports = function createRenderMiddleware(webpackConfig) {
-  const resolvable = new Resolvable();
-  const compiler = Object.assign(webpack(webpackConfig), {
-    outputFileSystem: new MemoryFS(),
+  const transpiler = exports.createTranspiler(webpackConfig);
+  let middleware, error;
+  transpiler.on('start', () => {
+    middleware = null;
+    error = null;
+  });
+  transpiler.on('success', (result) => {
+    middleware = result;
+    error = null;
   });
+  transpiler.on('error', (result) => {
+    middleware = null;
+    error = result;
+  });
+  const getMiddlewarePromise = () => {
+    if (error) {
+      return Promise.reject(error);
+    }
+    if (middleware) {
+      return Promise.resolve(middleware);
+    }
+    return new Promise((resolve) => {
+      transpiler.once('result', () => {
+        resolve(getMiddlewarePromise());
+      });
+    });
+  };
+  return function webpackRenderMiddleware(req, res, next) {
+    getMiddlewarePromise()
+      .then((middleware) => middleware(req, res, next))
+      .catch(next);
+  };
+};
+
+exports.createTranspiler = function createTranspiler(webpackConfig) {
+  const emitter = new EventEmitter();
+  const compiler = webpack(webpackConfig);
+  compiler.outputFileSystem = new MemoryFS();
+  compiler.hooks.watchRun.tap('untool-transpiler', () => emitter.emit('start'));
+  const emitResult = (...args) => {
+    emitter.emit(...args);
+    emitter.emit('result');
+  };
   const handleCompilation = (compileError, stats) => {
     if (compileError) {
-      resolvable.reject(compileError);
-    } else if (stats.hasErrors() || stats.hasWarnings()) {
-      const { errors, warnings } = stats.toJson();
-      resolvable.reject(
-        new Error(`Can't compile:\n${[...errors, ...warnings].join('\n')}`)
-      );
+      emitResult('error', compileError);
+    } else if (stats.hasErrors()) {
+      const { errors } = stats.toJson();
+      const details = errors
+        .map((str) => str.replace(/\033?\[[0-9]{1,2}m/g, ''))
```

Comment:
What are you doing here? Replacing color codes?

Reply:
Seems like you can set `colors: false` to the `stats.toJson()` method - https://webpack.js.org/configuration/stats/

Reply:
Yes, stripping color codes - I tried just using that setting, but it does not work the way we both thought. ;)